cpu/kinetis_common/uart.c

```diff
@@ -0,0 +1,323 @@
+ * @param uart          the UART device that triggered the ISR
+ */
+static inline void irq_handler(uart_t uartnum, UART_Type *uart);
+
+/**
+ * @brief Allocate memory to store the callback functions.
+ */
+static uart_conf_t config[UART_NUMOF];
+
+int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, uart_tx_cb_t tx_cb, void *arg)
+{
+    /* do basic initialization */
+    int res = uart_init_blocking(uart, baudrate);
+
+    if (res < 0) {
+        return res;
+    }
+
+    /* remember callback addresses */
+    config[uart].rx_cb = rx_cb;
+    config[uart].tx_cb = tx_cb;
+    config[uart].arg = arg;
+
+    /* enable receive interrupt */
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            NVIC_SetPriority(UART_0_IRQ_CHAN, UART_IRQ_PRIO);
+            NVIC_EnableIRQ(UART_0_IRQ_CHAN);
+            UART_0_DEV->C2 |= (1 << UART_C2_RIE_SHIFT);
+            break;
+#endif
+#if UART_1_EN
+
+        case UART_1:
+            NVIC_SetPriority(UART_1_IRQ_CHAN, UART_IRQ_PRIO);
+            NVIC_EnableIRQ(UART_1_IRQ_CHAN);
+            UART_1_DEV->C2 |= (1 << UART_C2_RIE_SHIFT);
+            break;
+#endif
+
+        default:
+            return -2;
+            break;
+    }
+
+    return 0;
+}
+
+int uart_init_blocking(uart_t uart, uint32_t baudrate)
+{
+    UART_Type *dev;
+    PORT_Type *port;
+    uint32_t clk;
+    uint16_t ubd;
+    uint16_t brfa;
+    uint8_t tx_pin = 0;
+    uint8_t rx_pin = 0;
+    uint8_t af;
+
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            dev = UART_0_DEV;
+            port = UART_0_PORT;
+            clk = UART_0_CLK;
+            tx_pin = UART_0_TX_PIN;
+            rx_pin = UART_0_RX_PIN;
+            af = UART_0_AF;
+            UART_0_PORT_CLKEN();
+            UART_0_CLKEN();
+            break;
+#endif
+#if UART_1_EN
+
+        case UART_1:
+            dev = UART_1_DEV;
+            port = UART_1_PORT;
+            clk = UART_1_CLK;
+            tx_pin = UART_1_TX_PIN;
+            rx_pin = UART_1_RX_PIN;
+            af = UART_1_AF;
+            UART_1_PORT_CLKEN();
+            UART_1_CLKEN();
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* configure RX and TX pins, set pin to use alternative function mode */
+    port->PCR[rx_pin] = PORT_PCR_MUX(af);
+    port->PCR[tx_pin] = PORT_PCR_MUX(af);
+
+    /* disable transmitter and receiver */
+    dev->C2 &= ~(1 << UART_C2_TE_SHIFT | 1 << UART_C2_RE_SHIFT);
+    /* set defaults, 8-bit mode, no parity */
+    dev->C1 = 0;
+
+    /* calculate baudrate */
+    ubd = (uint16_t)(clk / (baudrate * 16));
+    /* baudrate fine adjust - brfa */
+    brfa = (uint16_t)(((clk * 32) / (baudrate * 16)) - (ubd * 32));
+
+    /* set baudrate */
+    dev->BDH = (uint8_t)UART_BDH_SBR(ubd >> 8);
+    dev->BDL = (uint8_t)UART_BDL_SBR(ubd);
+
+    /* set brfa */
+    /* TODO: handle different uart types */
+    (void)brfa;
+    //dev->C4 = (uint8_t)UART_C4_BRFA(brfa);
```

Comment:
>Regarding UART: At least the k20 has two UART interrupts; status and error. Should both be >handled?

Some UART implementation (KL devices) have one for both.
Currently I have no idea. I am on the UART driver, but MCG Driver has come in between.